hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/PrepareSubCommand.java

```diff
@@ -0,0 +1,173 @@
+ * the License.
+ */
+
+package org.apache.hadoop.ozone.admin.om;
+
+import static org.apache.hadoop.ozone.OmUtils.getOmHostsFromConfig;
+import static org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.PrepareStatusResponse.PrepareStatus.PREPARE_COMPLETED;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.stream.Collectors;
+
+import org.apache.hadoop.hdds.cli.HddsVersionProvider;
+import org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.PrepareStatusResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.PrepareStatusResponse.PrepareStatus;
+
+import picocli.CommandLine;
+
+/**
+ * Handler of ozone admin om finalizeUpgrade command.
+ */
+@CommandLine.Command(
+    name = "prepare",
+    description = "Prepares Ozone Manager for upgrade/downgrade, by applying " +
+        "all pending transactions, taking a Ratis snapshot at the last " +
+        "transaction and purging all logs on each OM instance. The returned " +
+        "transaction #ID corresponds to the last transaction in the quorum in" +
+        " which the snapshot is taken.",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class
+)
+public class PrepareSubCommand implements Callable<Void> {
+
+  @CommandLine.ParentCommand
+  private OMAdmin parent;
+
+  @CommandLine.Option(
+      names = {"-id", "--service-id"},
+      description = "Ozone Manager Service ID",
+      required = true
+  )
+  private String omServiceId;
+
+  @CommandLine.Option(
+      names = {"-tawt", "--transaction-apply-wait-timeout"},
+      description = "Max time in SECONDS to wait for all transactions before" +
+          "the prepare request to be applied to the OM DB.",
+      defaultValue = "120",
+      hidden = true
+  )
+  private long txnApplyWaitTimeSeconds;
+
+  @CommandLine.Option(
+      names = {"-tact", "--transaction-apply-check-interval"},
+      description = "Time in SECONDS to wait between successive checks for " +
+          "all transactions to be applied to the OM DB.",
+      defaultValue = "5",
+      hidden = true
+  )
+  private long txnApplyCheckIntervalSeconds;
+
+  @CommandLine.Option(
+      names = {"-pct", "--prepare-check-interval"},
+      description = "Time in SECONDS to wait between successive checks for OM" +
+          " preparation.",
+      defaultValue = "10",
+      hidden = true
+  )
+  private long prepareCheckInterval;
+
+  @CommandLine.Option(
+      names = {"-pt", "--prepare-timeout"},
+      description = "Max time in SECONDS to wait for all OMs to be prepared",
+      defaultValue = "300",
+      hidden = true
+  )
+  private long prepareTimeOut;
+
+  @Override
+  public Void call() throws Exception {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 150-168 into a separate method.